libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,827 @@
+            .primary(args_.x)
+            .note("skipping invalid events")
+            .emit(dh);
+          consumed += xs.length();
+          continue;
+        }
+      }
+      if (args_.res) {
+        xs = floor(xs);
+      }
+      bucket* b = nullptr;
+      for (auto i = size_t{};
+           auto&& [idx, x] : detail::enumerate(xs.values())) {
+        const auto group_name = std::invoke([&]() -> std::string_view {
+          if (gs.array->IsNull(idx)) {
+            if (args_.group) {
+              diagnostic::warning("got group name `null`")
+                .primary(args_.group.value())
+                .note("using `\"null\"` instead")
+                .emit(dh);
+              return "null";
+            }
+            return "";
+          }
+          auto&& [it, _]
+            = gnames.emplace(value_at(string_type{}, *gs.array, idx));
+          return *it;
+        });
+        auto* newb = get_bucket(groups, plugins, x, group_name, s);
+        if (b != newb) {
+          if (b) {
+            for (auto&& instance : *b) {
+              instance->update(subslice(slice, consumed, consumed + i), s);
+            }
+          }
+          b = newb;
+          consumed += i;
+          i = 0;
+        }
+        ++i;
+      }
+      if (b) {
+        for (auto&& instance : *b) {
+          if (consumed != slice.rows()) {
+            instance->update(subslice(slice, consumed, slice.rows()), s);
+          }
+        }
+      }
+    }
+    if (groups.empty()) {
+      diagnostic::warning("chart_{} received no valid data",
+                          to_string(args_.ty))
+        .primary(args_.op_loc)
+        .emit(dh);
+      co_yield {};
+      co_return;
+    }
+    auto ynames = std::unordered_map<std::string, bool>{};
+    auto b = series_builder{};
+    const auto add_y = [&](std::string_view group, std::string_view y,
+                           bool valid) -> const std::string& {
+      auto yname = std::invoke([&] {
+        if (not args_.group) {
+          return std::string{y};
+        }
+        if (args_.y.size() == 1) {
+          return std::string{group};
+        }
+        return fmt::format("{}_{}", group, y);
+      });
+      auto [it, _] = ynames.try_emplace(std::move(yname), valid);
+      it->second &= valid;
+      return it->first;
+    };
+    const auto insert = [&](data x, const grouped_bucket& groups) {
+      auto r = b.record();
+      r.field(xpath).data(std::move(x));
+      for (auto&& [name, bucket] : groups) {
+        for (auto&& [y, instance] : detail::zip_equal(args_.y, bucket)) {
+          auto value = to_double(instance->get());
+          auto valid = validate_y(value, y.second.get_location(), dh);
+          r.field(add_y(name, y.first, valid)).data(std::move(value));
+        }
+      }
+    };
+    for (const data* prev = nullptr; const auto& [x, gb] : groups) {
+      if (args_.ty == chart_type::line and args_.res) {
```

Comment:
I remember that you said that the front-end doesn't support this yet, but area charts with `resolution` and gaps can be very misleading. Should we create an issue for this to follow-up on this later, create a warning in the docs, or what should we do?

Reply:
https://github.com/tenzir/issues/issues/2797